language/tools/move-build/src/source_package/manifest_parser.rs

```diff
@@ -0,0 +1,447 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::source_package::parsed_manifest as PM;
+use anyhow::{bail, format_err, Context, Result};
+use move_core_types::{account_address::AccountAddress, identifier::Identifier};
+use std::{
+    collections::{BTreeMap, BTreeSet},
+    path::PathBuf,
+};
+use toml::Value as TV;
+use url::Url;
+
+const EMPTY_ADDR_STR: &str = "_";
+
+const PACKAGE_NAME: &str = "package";
+const BUILD_NAME: &str = "build";
+const ADDRESSES_NAME: &str = "addresses";
+const DEV_ADDRESSES_NAME: &str = "dev_addresses";
+const DEPENDENCY_NAME: &str = "dependencies";
+const DEV_DEPENDENCY_NAME: &str = "dev_dependencies";
+
+const KNOWN_NAMES: &[&str] = &[
+    PACKAGE_NAME,
+    BUILD_NAME,
+    ADDRESSES_NAME,
+    DEV_ADDRESSES_NAME,
+    DEPENDENCY_NAME,
+    DEV_DEPENDENCY_NAME,
+];
+
+const REQUIRED_FIELDS: &[&str] = &[PACKAGE_NAME];
+
+pub fn parse_toml_and_unwrap_errors(manifest_string: String) -> Result<TV> {
+    match toml::from_str::<TV>(&manifest_string) {
+        Ok(toml_val) => Ok(toml_val),
+        Err(err) => Err(format_err!(
+            "Unable to parse as valid TOML.\nParsing error is:\n{}",
```

Comment:
I have a few thoughts on this error message:

- It would be nice to combine the 1st and 2nd lines into a single line.
- The 2nd line ("Parsing error is:") doesn't add much and could just be omitted.
- How much do we expect users to understand "valid TOML"? I expect we should document this format for "Move package manifests" and not rely on people understanding the TOML format.

I'd probably go with a single line saying something like: "Unable to parse Move package manifest:\n"